### ISSUE TITLE
implicit declaration of function 'opal_atomic_swap_ptr' is invalid in C99

### ISSUE THREAD

**Issue by krl52** [open-mpi/ompi#11137]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

tried 4.1.4 and 4.1.3

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

source tarball from https://www.open-mpi.org/software/ompi/v4.1/

### Please describe the system on which you are running

* Operating system/version: macOS Ventura 13.0.1
* Computer hardware: ?
* Network type: ?

-----------------------------

## Details of the problem

Please describe, in detail, the problem that you are having, including the behavior you expect to see, the actual behavior that you are seeing, steps to reproduce the problem, etc.  It is most helpful if you can attach a small program that a developer can use to reproduce your problem.
To install openMPI, I first ran: `./configure --prefix=$HOME/opt/openmpi`
and got the result: 
`configure: error: C and Fortran compilers are not link compatible. >> Can not continue.
`Solved this by changing to:
`./configure --prefix=$HOME/opt/openmpi CC="gcc -arch ${r_arch:=x86_64}" CXX="g++ -arch ${r_arch:=x86x64}" FC="gfortran -arch ${r_arch=x86_64}"
`
The end of the return is: 
```
config.status: creating opal/include/opal_config.h
config.status: creating ompi/include/mpi.h
config.status: creating oshmem/include/shmem.h
config.status: creating opal/mca/hwloc/hwloc201/hwloc/include/private/autogen/config.h
config.status: creating opal/mca/hwloc/hwloc201/hwloc/include/hwloc/autogen/config.h
config.status: creating ompi/mpiext/cuda/c/mpiext_cuda_c.h
config.status: executing depfiles commands
config.status: executing ompi/mca/osc/monitoring/osc_monitoring_template_gen.h commands
config.status: executing libtool commands

Open MPI configuration:
-----------------------
Version: 4.1.3
Build MPI C bindings: yes
Build MPI C++ bindings (deprecated): no
Build MPI Fortran bindings: mpif.h, use mpi, use mpi_f08
MPI Build Java bindings (experimental): no
Build Open SHMEM support: no (disabled)
Debug build: no
Platform file: (none)

Miscellaneous
-----------------------
CUDA support: no
HWLOC support: external
Libevent support: external
PMIx support: Internal
 
Transports
-----------------------
Cisco usNIC: no
Cray uGNI (Gemini/Aries): no
Intel Omnipath (PSM2): no
Intel TrueScale (PSM): no
Mellanox MXM: no
Open UCX: no
OpenFabrics OFI Libfabric: no
OpenFabrics Verbs: no
Portals4: no
Shared memory/copy in+copy out: yes
Shared memory/Linux CMA: no
Shared memory/Linux KNEM: no
Shared memory/XPMEM: no
TCP: yes
 
Resource Managers
-----------------------
Cray Alps: no
Grid Engine: no
LSF: no
Moab: no
Slurm: yes
ssh/rsh: yes
Torque: no
 
OMPIO File Systems
-----------------------
DDN Infinite Memory Engine: no
Generic Unix FS: yes
IBM Spectrum Scale/GPFS: no
Lustre: no
PVFS2/OrangeFS: no

```
Then when I run make all, the end of the output:

```
In file included from opal_datatype_pack.c:28:
In file included from ../../opal/datatype/opal_convertor_internal.h:21:
In file included from ../../opal/datatype/opal_convertor.h:35:
In file included from ../../opal/datatype/opal_datatype.h:41:
In file included from ../../opal/class/opal_object.h:126:
../../opal/threads/thread_usage.h:163:1: error: implicit declaration of function 'opal_atomic_swap_ptr' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
OPAL_THREAD_DEFINE_ATOMIC_SWAP(void *, intptr_t, ptr)
^
../../opal/threads/thread_usage.h:143:16: note: expanded from macro 'OPAL_THREAD_DEFINE_ATOMIC_SWAP'
        return opal_atomic_swap_ ## suffix ((volatile type *) ptr, newvalue); \
               ^
<scratch space>:37:1: note: expanded from here
opal_atomic_swap_ptr
^
../../opal/threads/thread_usage.h:163:1: note: did you mean 'opal_thread_swap_ptr'?
../../opal/threads/thread_usage.h:143:16: note: expanded from macro 'OPAL_THREAD_DEFINE_ATOMIC_SWAP'
        return opal_atomic_swap_ ## suffix ((volatile type *) ptr, newvalue); \
               ^
<scratch space>:37:1: note: expanded from here
opal_atomic_swap_ptr
^
../../opal/threads/thread_usage.h:163:1: note: 'opal_thread_swap_ptr' declared here
../../opal/threads/thread_usage.h:140:20: note: expanded from macro 'OPAL_THREAD_DEFINE_ATOMIC_SWAP'
static inline type opal_thread_swap_ ## suffix (volatile addr_type *ptr, type newvalue) \
                   ^
<scratch space>:36:1: note: expanded from here
opal_thread_swap_ptr
^
In file included from opal_datatype_pack.c:28:
In file included from ../../opal/datatype/opal_convertor_internal.h:21:
In file included from ../../opal/datatype/opal_convertor.h:35:
In file included from ../../opal/datatype/opal_datatype.h:41:
In file included from ../../opal/class/opal_object.h:126:
../../opal/threads/thread_usage.h:163:1: warning: incompatible integer to pointer conversion returning 'int' from a function with result type 'void *' [-Wint-conversion]
OPAL_THREAD_DEFINE_ATOMIC_SWAP(void *, intptr_t, ptr)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../opal/threads/thread_usage.h:143:16: note: expanded from macro 'OPAL_THREAD_DEFINE_ATOMIC_SWAP'
        return opal_atomic_swap_ ## suffix ((volatile type *) ptr, newvalue); \
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:37:1: note: expanded from here
opal_atomic_swap_ptr
^
1 warning and 18 errors generated.
make[2]: *** [libdatatype_reliable_la-opal_datatype_pack.lo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1

```

Is it related to #11114?

**Note**: If you include verbatim output (or a code block), please use a [GitHub Markdown](https://help.github.com/articles/creating-and-highlighting-code-blocks/) code block like below:
```shell
shell$ mpirun -n 2 ./hello_world
```

**Comment by devreal** [comment]
@krl52 Thanks for the report. It seems strange to me that gcc and gfortran are not link compatible. I assume that configure gets confused when you force the architecture. Which versions are you using? How were they installed? My gcc does not support the `-arch` flag (on Linux) but maybe that is a mac OS specialty?

**Comment by jsquyres** [comment]
I notice you have a mismatch in your architecture values between C/C++/Fortran (C++ is different).  Is that a typo here in the github issue, or was that intentional?

**Comment by krl52** [comment]
@jsquyres Yes, you're right, thanks for pointing that out! I tried fixing the typo after running autoreconf -I but unfortunately that didn't fix the issue.

@devreal I installed with Brew. output of gcc --version:
```
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

```

output of gfortran --version
```
GNU Fortran (Homebrew GCC 12.2.0) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

**Comment by jsquyres** [comment]
You should not need to run autoreconf -- Open MPI tarballs are fully bootstrapped.

It may be safer to `rm -rf` your build tree, and -- if necessary -- re-expand the tarball (to get a clean, fresh environment), then run `configure` again with the correct values and see if that fixes the problem.

**Comment by bosilca** [comment]
I am to reproduce this on my OSX. I tried all combinations gcc (12) and gfortran (12), clang (Apple version), gfortran (12), clang (14) and gfortran (12) and they all work. Maybe the difference is that I use MacPort to install these packages, or maybe that I have the M1 CPU. Please post the config.log file of the initial test (without the manually specified architectures).

**Comment by krl52** [comment]
@jsquyres I gave that a try, but unfortunately am still running into the same issue. @bosilca I do have the M2 CPU. I attached the config.log file for the original gcc and gfortran linker error.
[config.log](https://github.com/open-mpi/ompi/files/10136685/config.log)

**Comment by bosilca** [comment]
```
configure:6560: gcc --version >&5
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.1.0
```

```
configure:43812: gfortran -v >&5
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/12.2.0/bin/../libexec/gcc/x86_64-apple-darwin22/12/lto-wrapper
Target: x86_64-apple-darwin22
```

I think you are using an old version of gfortran, that can only generate code for x86_64. You might want to rebuild your brew, or make sure you are using the correct compilers.

**Comment by krl52** [comment]
OK, I have solved the issue. I ran `brew doctor` which came up with a number of warnings. I started working through them and eventually got: `Your Homebrew's prefix is not /opt/homebrew`. Uninstalled and re-installed Homebrew and ran `brew install gfortran`. Now it works!! Thanks for your help!

I have a suspicion about what may have happened. My old computer broke down last week and I replaced it. I was able to transfer all of the content from old to new computer. Old computer had an intel chip, new one has the M2. According to the Homebrew doc, the default prefix is /usr/local for Mac on Intel and /opt/homebrew for Mac on Apple Silicon/ARM.
